### ISSUE TITLE
Positional Encoding

### ISSUE THREAD

**Issue by hila-chefer** [lucidrains/vit-pytorch#11]
Hi @lucidrains ! 
Is there any specific reason that you define the positional encodings to be a learnable parameter that is randomly initialized?
As I understand from the paper they use the “standard” 1D positional encoding and I assumed it was the same as in the Transformer Encoder as demonstrated in this link:
https://pytorch.org/tutorials/beginner/transformer_tutorial.html

Thanks!

**Comment by lucidrains** [comment]
@hila-chefer yup, that is what I have, I'm using a standard 1d positional encoding

**Comment by lucidrains** [comment]
2d would look something like https://github.com/lucidrains/axial-positional-embedding, and you are free to try it, but it didn't make a difference in their experiments

**Comment by hila-chefer** [comment]
Sorry for rehashing this but this is the positional encoding function from the Transformer implementation that I attached (using the sin + cos functions), and I couldn’t find that in this implementation:

class PositionalEncoding(nn.Module):

    def __init__(self, d_model, dropout=0.1, max_len=5000):
        super(PositionalEncoding, self).__init__()
        self.dropout = nn.Dropout(p=dropout)

        pe = torch.zeros(max_len, d_model)
        position = torch.arange(0, max_len, dtype=torch.float).unsqueeze(1)
        div_term = torch.exp(torch.arange(0, d_model, 2).float() * (-math.log(10000.0) / d_model))
        pe[:, 0::2] = torch.sin(position * div_term)
        pe[:, 1::2] = torch.cos(position * div_term)
        pe = pe.unsqueeze(0).transpose(0, 1)
        self.register_buffer('pe', pe)

    def forward(self, x):
        x = x + self.pe[:x.size(0), :]
        return self.dropout(x)

Am I missing something?

Thanks again!

**Comment by lucidrains** [comment]
@hila-chefer Yup, that isn't what they used in the paper, they used a 1d learned positional encoding. What you showed me is called a fixed sinusoidal positional encoding.

**Comment by hila-chefer** [comment]
Ok, thanks for clarifying that! Wasn’t aware of that.

**Comment by lucidrains** [comment]
@hila-chefer 

quick primer

1. fixed sinusoidal encoding - no learned parameters
2. absolute positional encoding - 1d learned encoding
3. axial positional encoding - 2d learned encoding
